### PR TITLE
BUG: fix computation of hyp2f1 near integer values

### LINKED ISSUE #6491: Weird behaviour of special.hyp2f1 with numpy arrays
Hello,

I encountered this strange behaviour when using special.hyp2f1
`In [53]: b = np.arange(1.95, 2.35, 0.1); b
Out[53]: array([ 1.95,  2.05,  2.15,  2.25,  2.35])

In [54]: scsp.hyp2f1(0.25,b,1.25, -10)
Out[54]: array([ 0.47118224,  0.46334121,  0.45605508,        -inf,  0.44291621])

In [55]: b = np.arange(2.15, 2.35, 0.1); b
Out[55]: array([ 2.15,  2.25,  2.35])

In [56]: scsp.hyp2f1(0.25,b,1.25, -10)
Out[56]: array([ 0.45605508,  0.44926403,  0.44291621])
`

My Scipy version is 0.17.1. Numpy version 1.11.1.

Do you have an idea on how to wokaround this ?

### PR DESCRIPTION
**Pull request by ghost**
#### Reference issue
Closes gh-6491

#### What does this implement/fix?
Constrain input parameters, and certain differences of input parameters, to be exactly integers whenever the values are tested to be nearly integers (within EPS).  This prevents incorrect results with certain inputs.  Unit test behavior is unchanged.

### CODE REVIEW

**Comment by steppi** [comment]
@femler0 Thanks for submitting this PR.  I should be able to take a look over the weekend. We've been working on fixing the issues with `hyp2f1` for complex inputs but haven't touched the float/double cases yet. SciPy's special functions can dispatch to different implementations based on type, and at the moment we use cephes for real input and [our own implemention](https://github.com/scipy/scipy/blob/main/scipy/special/_hyp2f1.pxd) for complex input. 

One thing I can say now is that we should add some new tests which fail on main and pass on this branch before a fix goes in. The relevant test code is [here](https://github.com/scipy/scipy/blob/9da5381f1d75c6a61261009b5517ba54719d87aa/scipy/special/tests/test_hyp2f1.py#L377-L530). It would be great if you can add the cases from gh-6491 in the relevant places. You'll need to change the condition [here](https://github.com/scipy/scipy/blob/9da5381f1d75c6a61261009b5517ba54719d87aa/scipy/special/tests/test_hyp2f1.py#L464) in order to allow small perturbations in `a` and `b` for the `a` or `b` a negative int tests.

**Comment by femler0** [comment]
@steppi I added a test case and verified it fails on main and succeeds on the updated code.

**Comment by steppi** [comment]
Hi @femler0, it looks like you've accidentally undone the changes from your first commit in your second commit. If you're using the git command line ui, one way to fix this is to run

````git RESET HEAD~````

to undo the most recent commit, but leave your files unchanged. Then run

```git restore <path to>/special/cephes/hyp2f1.c```

to bring `hyp2f1.c` back to the state from your first commit. Then you can stage and commit `test_hyp2f1.py` again. After that, you will need to force push back to this branch by running `git push` with the `-f` option. Let me know if you have any questions. I can do these steps on my local copy of your branch to fix things up if you like. 

I have some review suggestions to make, but first let's get your branch back into shape.

**Comment by femler0** [comment]
Sorry about that, I had been going back and forth between main and the feature branch to verify not fixed vs. fixed and somehow screwed up.  I use git every day, but am new to github development/forks/pull requests/etc.

Anyway, I fixed the commit and also rebased on top of the latest upstream/main.

**Comment by steppi** [comment]
> Sorry about that, I had been going back and forth between main and the feature branch to verify not fixed vs. fixed and somehow screwed up. I use git every day, but am new to github development/forks/pull requests/etc.
> 
> Anyway, I fixed the commit and also rebased on top of the latest upstream/main.

No worries @femler0. That can happen to anyone, especially when working on several branches at once. Thanks for fixing things up.

I think you're approach is sound, but based on some quick numerical investigations, it looks like we'll also want to be more strict when determining whether one of `a`, `b`, `c - a`, or `c - b` is a negative integer.

Using the script from the details section below, I generated the following table:

<details>

```python
import argparse
import itertools as it
import numpy as np
import pandas as pd
import scipy.special as sc

from mpmath import mp


parser = argparse.ArgumentParser()
parser.add_argument("epsilon", type=float)
args = parser.parse_args()

# Seed generated with np.random.SeedSequence.entropy
rng = np.random.default_rng(
    171674233887121422766798348511263005334
)


X = np.arange(-10, 1, 0.1)
# Negative int value for a peturbed by small amount
A = np.arange(-10, 0) + args.epsilon
B = 2**np.arange(0, 5) + rng.normal(scale=0.1, size=5)
C = 2**np.arange(0, 5) + rng.normal(scale=0.1, size=5)
B = np.hstack([-B, B])
C = np.hstack([-C, C])

errors1 = []
errors2 = []
for a, b, c in it.product(A, B, C):
    observed1 = sc.hyp2f1(a, b, c, X)
    observed2 = sc.hyp2f1(a, b, c, X.astype(complex))
    expected = []
    for x in X:
        expected.append(float(mp.hyp2f1(a, b, c, x)))
    expected = np.asarray(expected)
    rel_error1 = abs(expected - observed1) / abs(expected)
    rel_error2 = abs(expected - observed2) / abs(expected)
    errors1.append(rel_error1)
    errors2.append(rel_error2)

errors1 = np.hstack(errors1)
errors2 = np.hstack(errors2)

quantiles1 = np.quantile(
    rel_error1, [0.25, 0.5, 0.75, 0.9, 1]
)

quantiles2 = np.quantile(
    rel_error2, [0.25, 0.5, 0.75, 0.9, 1]
)

rows = [quantiles1, quantiles2]
df = pd.DataFrame(rows)
df.index = ["This PR", "Strict"]
df.columns = [
    "25%", "50%", "75%", "90%", "max"
]
print(df.to_markdown(mode="github"))
````
</details>

|         |         25% |         50% |         75% |         90% |         max |
|:--------|------------:|------------:|------------:|------------:|------------:|
| This PR | 1.12378e-14 | 1.70925e-14 | 2.09497e-14 | 2.26192e-14 | 2.37308e-14 |
| Strict  | 1.23521e-16 | 1.90563e-16 | 3.45869e-16 | 5.32193e-16 | 1.42505e-15 |

This gives various percentiles for relative error over 110,000 test cases where `a` is equal to a negative integer plus `0.99e-14`. The first row uses the implementation from this PR, and the second uses the `hyp2f1` implementation for complex numbers, which requires the exact condition `a == round(a)` to consider `a` to be an integer.

We're losing precision by considering `hyp2f1` to be constant in `a` when `a` is near a negative integer. It seems like the standard series and transformations are handling these cases fine though. I should have some time to finish my review tomorrow evening.

**Comment by femler0** [comment]
That's good analysis.  My solution was definitely a band-aid solution, rather than a comprehensive one.  I think the hierarchy of integer vs. non-integer tests, where the test is performed at different levels, is the source of the current bug.  I could go back and reason about it more carefully, but I think the changes would end up being more extensive.  The reason I didn't attempt that solution was I thought the risk vs. reward was too great, especially for my first scipy contribution.  But loss of precision is a realized risk, so I miscalculated.

Would it be better to wait for the complex-valued case to be completed and then use whatever decision tree that is using in the real-valued case as well?

**Comment by steppi** [comment]
> Would it be better to wait for the complex-valued case to be completed and then use whatever decision tree that is using in the real-valued case as well?

No need to wait. Let's get a fix in. It'll be good to have two sets of eyes on it. (The logic for this is already settled for complex numbers, but we may want to do things a little differently here). I'll go through now and check if there would be any unintended impact if we set lower thresholds to consider `a`, `b`, `c`, etc... to be integers.

**Comment by steppi** [comment]
In principle, it seems like we might be able to get away with just changing to `define EPS 1.0e-15`, on top of the changes you've already made. We'd need to test this very thoroughly though. It's not clear to me why this threshold was chosen.
There's a limiting formula used when `x > 0.9` and `c - a - b` is within `EPS` of 0, which will not work when one of `a` or `b` is a negative integer.

```c
	    /* Psi function expansion, AMS55 #15.3.10, #15.3.11, #15.3.12
	     *
	     * Although AMS55 does not explicitly state it, this expansion fails
	     * for negative integer a or b, since the psi and Gamma functions
	     * involved have poles.
	     */
```

and I suspect this might be involved in the choice of cutoff, but I didn't see any issues when setting EPS=1.0e-15` for the handful of relevant cases I checked.
We have a somewhat comprehensive benchmark suite for `hyp2f1` for complex values [here] that could be adapted for real values. 

Adjusting `EPS` seems fairly involved. If you're interested, we can follow up with that later. I think it might be OK just to apply the band-aid solution for now. I'll look into that a little more the next time I have a chance, probably over the weekend.

**Comment by steppi** [comment]
@femler0

Let's try to get this fix in, but first let's test it thoroughly. There's a [script](https://github.com/scipy/scipy/blob/main/scipy/special/_precompute/hyp2f1_data.py) in `scipy/special/_precompute` which exhaustively tests `hyp2f1` for complex numbers. This takes way to long to run to be part of the ordinary test suite. We should make a similar script for real values. We can change the directory structure to have `scipy/special/_precompute/hyp2f1`, and have `hyp2f1_data_complex.py` and `hyp2f1_data_real.py` within this parent folder, with the former being the existing script.

In `hyp2f1_data_real.py` let's add a set of test cases where one or more of `a`, `b`, `c - a`, `c - b` either equal to a negative integer, less than `1e-13` from a negative integer, or just more than `1e-13` from a negative integer. When one of `a` or`b` is a negative int, they should vary across orders of magnitude. At or very near powers of 2 from 2 to 128 sounds reasonable. For `c - a` or `c - a` a negative int, I think it's OK to just have a few different values. `x` can take values in a grid from -2 to just before 1.

This will be a lot of test cases. We can select maybe 5-10 of them randomly to include in `scipy/special/test/test_hyp2f1.py`.

Would you be interested in working on this?

**Comment by femler0** [comment]
@steppi 

TL;DR  version:  Yes, I can write up a test script and execute it.  I haven't looked at your example in detail yet.  Around negative integers it seems like using [this function](https://numpy.org/doc/stable/reference/generated/numpy.nextafter.html) could be useful.  My suspicion is that I will find many other cases of paradoxical behavior.  At that point, it might make the most sense to examine the handful of "endpoints" in the code (i.e., where the recursion ends) individually to see where they fail.  I haven't lost interest in this yet (though I've come close several times :smiley: ).

Long version:

I have tried several times to move stuff around in hyp2f1 in an effort to unravel how it works and what needs to be fixed.  The last time I tried (yesterday) was most successful.  I put some modified code [here](https://github.com/femler0/scipy/tree/whatever).  Note that it is WIP and I'm _not_ proposing using it.  But I think I was on the right track in making the web of transformations easier to understand (I was using the Pearson, et al reference mentioned in the complex-valued case as a guide).

In the modified code I had taken out the within EPS test for negative integers and the way I had rearranged things, the result for the test case we added was a seg fault (due to infinite recursion) rather than the paradoxical evaluation to infinity.  

Looking at the call stack:
```
#0  0x00007fffe107dd9a in cephes_hyp2f1 (a=2.2499999999999001, b=1, c=2.9999999999999001, x=0.090909090909090939) at scipy/special/cephes/hyp2f1.c:90
#1  0x00007fffe107e822 in cephes_hyp2f1 (a=2.2499999999999001, b=1, c=1.25, x=0.90909090909090906) at scipy/special/cephes/hyp2f1.c:221
#2  0x00007fffe107e822 in cephes_hyp2f1 (a=2.2499999999999001, b=1, c=2.9999999999999001, x=0.090909090909090939) at scipy/special/cephes/hyp2f1.c:221
#3  0x00007fffe107e822 in cephes_hyp2f1 (a=2.2499999999999001, b=1, c=1.25, x=0.90909090909090906) at scipy/special/cephes/hyp2f1.c:221
#4  0x00007fffe107e822 in cephes_hyp2f1 (a=2.2499999999999001, b=1, c=2.9999999999999001, x=0.090909090909090939) at scipy/special/cephes/hyp2f1.c:221
#5  0x00007fffe107e822 in cephes_hyp2f1 (a=2.2499999999999001, b=1, c=1.25, x=0.90909090909090906) at scipy/special/cephes/hyp2f1.c:221
#6  0x00007fffe107e822 in cephes_hyp2f1 (a=2.2499999999999001, b=1, c=2.9999999999999001, x=0.090909090909090939) at scipy/special/cephes/hyp2f1.c:221
#7  0x00007fffe107e822 in cephes_hyp2f1 (a=2.2499999999999001, b=1, c=1.25, x=0.90909090909090906) at scipy/special/cephes/hyp2f1.c:221
```
the code is both trying to make `a` a negative integer and making `x < 0.9`.  In the original code, I think the issue is that it was deciding to evaluate the x=0.90909... case but then it didn't converge so the result was infinity.  I _think_ the approach I added (adding error to the input `a` or `b` by rounding so that a polynomial can be evaluated), is an OK one, but the conditions for applying that fix will need to be examined in detail.  It would be great if there were theoretical results on what a small change to the input parameters does to the output result, but I'm guessing those don't exist.

I've gone through a good portion of AMS55 and the Pearson, Olver, Porter reference.  Seems like there's no end to the complexity.  I think for this bug fix, I'd like to draw the line at not writing new "endpoints" in the code, but only change the conditions in which those endpoints are reached and maximize accuracy and sanity within those constraints.

Regardless, the first step is to write the test suite you mentioned, using the existing one as a baseline.  Sounds like the results it generates are too big to share, so I'll share the test suite code once complete.

**Comment by steppi** [comment]
> TL;DR version: Yes, I can write up a test script and execute it

Great! I was determined to clean up the issues with `hyp2f1` but ended up taking on a more demanding job and had a lot of other things going on. If you're set on this, I'd be happy to help review your PRs. The key thing is to only work on a little bit at a time, and test things very thoroughly.

> It would be great if there were theoretical results on what a small change to the input parameters does to the output result, but I'm guessing those don't exist.

The topic you're interested in here is [Backwards error analysis](https://en.wikipedia.org/wiki/Error_analysis_(mathematics)#Backward_error_analysis), if you weren't aware of the name. It would be helpful to know the partial derivatives with respect to `a` and `b`. There unfortunately aren't closed forms for these just somewhat messy series expansions. The ones from [this link](https://math.stackexchange.com/a/4052952/7323) are valid for $-1 < z < 1$, and they would get even messier outside this range because we'd have to differentiate through the standard transformations. In any case, tests based on numerical calculations should be sufficient here.


> Regardless, the first step is to write the test suite you mentioned, using the existing one as a baseline. Sounds like the results it generates are too big to share, so I'll share the test suite code once complete.

Right, that's the first step and it's what I did first when working on the complex case. Once you get a similar test suite working, feel free to submit a PR for it and I can review.

**Comment by femler0** [comment]
OK, thanks.  It's going to be best if I don't make future contributions.  I didn't really "study" any other implementations, more like "glanced."  I'm confident there's not a true issue there, though I agree the appearance of impropriety is also an issue.  The larger issue is that I don't have the right kind of mind for this type of work.  I tend to obsess about problems like this and spend an unhealthy amount of time investigating things that aren't helpful to the project or my mental state.  I'm unable to focus on specific issues because of this, so I'm not a good collaborator.  I wanted to give back, but it's just not a good idea for anybody.

Wrapping up:
Re: script and test case I don't have much more to add  beyond what's in the linked branch.  I went back and made a "better" version that centered around integer-valued cases rather than the random number generation in the linked branch but it ended up taking inordinate amounts of time to run b/c mpmath struggles at near-integer cases.

Best wishes.

**Comment by steppi** [comment]
> OK, thanks. It's going to be best if I don't make future contributions. I didn't really "study" any other implementations, more like "glanced." I'm confident there's not a true issue there, though I agree the appearance of impropriety is also an issue. The larger issue is that I don't have the right kind of mind for this type of work. I tend to obsess about problems like this and spend an unhealthy amount of time investigating things that aren't helpful to the project or my mental state. I'm unable to focus on specific issues because of this, so I'm not a good collaborator. I wanted to give back, but it's just not a good idea for anybody.
> 
> Wrapping up: Re: script and test case I don't have much more to add beyond what's in the linked branch. I went back and made a "better" version that centered around integer-valued cases rather than the random number generation in the linked branch but it ended up taking inordinate amounts of time to run b/c mpmath struggles at near-integer cases.
> 
> Best wishes.

I’m sorry to hear that @femler0. I don’t think the results of your investigations were unhelpful. You’ve provided valuable info in the comments above. Work on `special` is slow at the moment, due to limited available time from maintainers and contributors with relevant expertise. I can see how slow and cautious development timeline could be tough for someone who tends to obsess over problems. You should definitely put your mental state first. 

I plan to create an omnibus issue with a checklist of hypergeometric function issues, where hypergeometric function development could be discussed. Your input would be appreciated there if you have more thoughts to share. I’ll give you a ping after I create it.

I’m actually unsure to what extent your ability to contribute could be impacted by glancing at GPL’d implementations. If you change your mind about contributing I would defer to a more experienced maintainer in that case.

Thanks again for your PR and the suggestions above.
